utils/misc.ml

```diff
@@ -222,6 +222,17 @@ module Stdlib = struct
         Some (Array.map (function None -> raise_notrace Exit | Some x -> x) a)
       with
       | Exit -> None
+
+    let equal eq_elt l1 l2 =
+      (* Basically inlines [Array.for_all2] to avoid the [raise] *)
+      let n = Array.length l1 in
+      Int.equal n (Array.length l2) &&
+      let rec loop i =
+        Int.equal i n ||
+        (eq_elt (Array.unsafe_get l1 i) (Array.unsafe_get l2 i) &&
+         loop (succ i))
```

Comment:
This isn't tail-recursive. :(

(Unless `||` and `&&` are. Which they might be. But the docs don't seem to guarantee it, so I wouldn't rely on it.)

Reply:
I'm 80% sure they are, having read the compiler source, but might be worth tweaking for clarity regardless